### PR TITLE
Experimental support for the new error format

### PR DESCRIPTION
**Pull request by Thinkofname**
This is mostly a result of me playing around with the new json formatted error messages but I decided to PR it to see what was thought of it.

It requires the latest dev build of Sublime Text 3 so I don't really expect this to be pulled as it stands.

This uses the phantom system to display error messages and hints inline with the code, i'm not sure how much I like this system but it seemed to be the nicest way I could think of displaying them.

I don't really do python so i'm sorry about the code

### CODE REVIEW

**Comment by dten** [comment]
I'm trying to try this out but where is expecting `rust_syntax_checking` to be enabled?

File: plugin.py [hunk 1]
```diff
@@ -31,55 +33,65 @@ def is_event_on_gutter(view, event):
...
-                msg = result.group(3).decode('utf-8')
-                if (fileName == view_filename and line):
-                    self.errors[line] = msg
-                    self.draw_dots_to_screen(view, int(line))
-                else:
-                    view.erase_regions('buildError')
-
+            view_filename = view.file_name()
+
+            for line in output[1].split(os.linesep):
```
Comment (dten) [inline plugin.py]: hard to change this line to the below to make it work. i'm using windows 10 + rustup installed rust windows-gnu nightly
`for line in output[1].split('\n'):`

**Comment by Thinkofname** [comment]
@dten I have a `Rust.sublime-settings` in `.config/sublime-text-3/Packages/User` (Preferences -> Browse packages) with 
```json

	// Enable the syntax checking plugin, which will highlight any errors during build
	"rust_syntax_checking": true
}
```
in it. 
I'm not sure why it doesn't show in package settings but it didn't show before this PR either. I haven't really worked with sublime's api before to be able to fix it

File: plugin.py (hunk 1 above)
Comment (Thinkofname) [inline plugin.py]: Fixed, thanks

**Comment by dten** [comment]
it's pretty sweet :+1: only thing is I currently get the cargo errors in-line _and_ your errors

**Comment by dten** [comment]
bit of poking around shows a file called `Main.sublime-menu` with something like this in it makes the settings appear in the menu
`[
    {
        "caption": "Preferences",
        "mnemonic": "n",
        "id": "preferences",
        "children":
        [
            {
                "caption": "Package Settings",
                "mnemonic": "P",
                "id": "package-settings",
                "children":
                [
                    {
                        "caption": "Rust",
                        "children":
                        [
                            {
                                "command": "open_file",
                                "args": {"file": "${packages}/Rust/Rust.sublime-settings"},
                                "caption": "Settings – Default"
                            },
                            {
                                "command": "open_file",
                                "args": {"file": "${packages}/User/Rust.sublime-settings"},
                                "caption": "Settings – User"
                            }
                        ]
                    }
                ]
            }
        ]
    }
]
`

**Comment by Thinkofname** [comment]
Seems like something that should be its own PR instead of adding it to this one. (Unless someone thinks it's a good idea)

**Comment by brson** [comment]
cc @steveklabnik 

I don't really have any understanding of sublime text, but I'll happily merge if someone says it's the right thing to do.

**Comment by brson** [comment]
cc @petrochenkov @Jayflux looks like you two have contributed recently. Maybe have opinions.

**Comment by Jayflux** [comment]
> It requires the latest dev build of Sublime Text 3 so I don't really expect this to be pulled as it stands.
@Thinkofname I don't think thats a problem, using the API you can detect 3 or above, and run it if its 3, and don't run it if its not.
See example here:
https://github.com/Jayflux/sublime-rust/blob/improvement/syntaxChecking/syntaxCheckerPlugin.py#L8

> I'm not sure why it doesn't show in package settings but it didn't show before this PR either. I haven't really worked with sublime's api before to be able to fix it

I never got that far to adding it, thanks for adding it though!

I would actually prefer this to go in, instead of my PR, as this seems more complete

**Comment by Thinkofname** [comment]
Added the package settings menu item using @dten's suggestion and added the version checking from @Jayflux, thanks.

I also cleaned things up a bit.

**Comment by Jayflux** [comment]
@Thinkofname just tested this branch, its working nicely!
testingrust

@dten can we switch this feature on by default? or would you rather users add a setting to have it.
Its a nice feature but not sure anyone would know its there

**Comment by dten** [comment]
My only issue is I currently get both errors when using building with cargo
Have no opinion if it should be on by default :+1: 
image

**Comment by Jayflux** [comment]
@dten  could you walk me through step by step how you got that to happen so i can re-create the problem?
Also are you on IRC?

**Comment by Thinkofname** [comment]
@dten Guessing thats using the build system feature of sublime (Ctrl + B)? 

Thats a plugin built-in to sublime (seperate from this) which uses the `file_regex` of the `.sublime-build` files we provide to place them. Removing that would unfortunately break using F4 to jump to the errors

**Comment by dten** [comment]
it was using Ctrl+B
I guess it should be possible to provide a build system programmatically that is the equiv of this on-save action that runs full cargo and picks the errors out like this ? Something to come back to I think

**Comment by Jayflux** [comment]
@Thinkofname  could you push a commit on this PR setting rust_syntax_checking  to true?

**Comment by Thinkofname** [comment]
@Jayflux sorry about the delay, done

**Comment by Jayflux** [comment]
@brson  @steveklabnik @dten 
I think this is ready to go in

**Comment by brson** [comment]
Thanks @Thinkofname and @Jayflux.

**Pull request merged**